File: Tyche.iOS/Session/Sources/Session/Authentication/Application/SignInWithEmailLinkUseCase.swift
import Core

public class SignInWithEmailLinkUseCase {
    let authenticationRepository: AuthenticationRepository
    let accountStorage: AccountStorage
    
    init(authenticationRepository: AuthenticationRepository, accountStorage: AccountStorage) {
        self.authenticationRepository = authenticationRepository
        self.accountStorage = accountStorage
    }
    
    public func execute(email: Email, emailLink: String) async -> Result<AccountBundle, Error> {
        let signInResult = await authenticationRepository.signInWithEmailLink(email: email.value, emailLink: emailLink)
        guard case .success(let externalAccountId) = signInResult else {
            return Result.failure(signInResult.errorOrNil()!)
        }
        
        let linkAccountResult = await authenticationRepository.linkAccount(
            accountLink: AccountLink(email: email, externalAccountId: externalAccountId)
        )
        guard case .success(let accountBundle) = linkAccountResult else {
            return Result.failure(linkAccountResult.errorOrNil()!)
        }

        do {
            try await accountStorage.store(accountBundle: accountBundle)
        } catch {
            return Result.failure(error)
        }
        
        return Result.success(accountBundle)
    }
}

File: Android/session/src/main/java/com/felipearpa/tyche/session/authentication/application/SignInWithEmailLinkUseCase.kt
package com.felipearpa.tyche.session.authentication.application

import com.felipearpa.tyche.core.type.Email
import com.felipearpa.tyche.session.AccountBundle
import com.felipearpa.tyche.session.AccountStorage
import com.felipearpa.tyche.session.authentication.domain.AccountLink
import com.felipearpa.tyche.session.authentication.domain.AuthenticationRepository

class SignInWithEmailLinkUseCase(
    private val authenticationRepository: AuthenticationRepository,
    private val accountStorage: AccountStorage
) {
    suspend fun execute(email: Email, emailLink: String): Result<AccountBundle> {
        val externalAccountId =
            authenticationRepository.signInWithEmailLink(email = email.value, emailLink = emailLink)
                .onFailure { exception -> return Result.failure(exception) }
                .getOrNull()!!

        val accountBundle = authenticationRepository.linkAccount(
            accountLink = AccountLink(
                email = email,
                externalAccountId = externalAccountId
            )
        )
            .onFailure { exception -> return Result.failure(exception) }
            .getOrNull()!!

        accountStorage.store(accountBundle = accountBundle)

        return Result.success(accountBundle)
    }
}